design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,185 @@
+# Updating OSDs in parallel
+**Targeted for v1.6**
+
+## Background
+In clusters with large numbers of OSDs, it can take a very long time to update all of the OSDs. This
+occurs on updates of Rook and Ceph both for major as well as the most minor updates. To better
+support large clusters, Rook should be able to update (and upgrade) multiple OSDs in parallel.
+
+
+## High-level requirements
+
+### Ability to set a maximumum number of OSDs to be updated simultaneously
+In the worst (but unlikely) case, all OSDs which are updated for a given parallel update operation
+might fail to come back online after they are updated. Users may wish to limit the number of OSDs
+updated in parallel in order to avoid too many OSDs failing in this way.
+
+### Cluster growth takes precedence over updates
+Adding new OSDs to a cluster should occur as quickly as possible. This allows users to make use of
+newly added storage as quickly as possible, which they may need for critical applications using the
+underlying Ceph storage. In some degraded cases, adding new storage may be necessary in order to
+allow currently-running Ceph OSDs to be updated without experiencing storage cluster downtime.
+
+This does not necessarily mean that adding new OSDs needs to happen before updates.
+
+This prioritization might delay updates significantly since adding OSDs not only adds capacity to
+the Ceph cluster but also necessitates data rebalancing. Rebalancing generates data movement which
+needs to settle for updates to be able to proceed.
+
+### OSD updates should not starve other resources of updates
+For Ceph cluster with huge numbers of OSDs, Rook's process to update OSDs should not starve other
+resources out of the opportunity to get configuration updates.
+
+
+## Technical implementation details
+
+### Parallelizing OSD creation and updates
+The `ceph osd ok-to-stop` results are only valid while there is no data movement within the Ceph 
+cluster. Adding new OSDs to the cluster introduces data movement, so Rook should only choose to 
+either add new OSDs or update a batch of existing OSDs at any given time. Rook should not create 
+new OSDs in a parallel thread to updating existing OSDs.
+
+Furthermore, Rook should wait for newly created OSDs to be recognized by Ceph before continuing to
+update OSD Deployments to ensure the results of `ceph osd ok-to-stop` are not invalidated by new
+OSD Pods coming online after the command returns.
```

Comment:
@leseb does this make sense?

Reply:
I'm not sure this will be a common scenario. Users will generally either be updating, or expanding their cluster, but not likely both at the same time. It seems like `ok-to-stop` will still handle this case of doing it at the same time, so we wouldn't need special consideration for this scenario, right?

Reply:
Even if it's not common, it will happen. It's my understanding that if we don't protect from the scenario, a large failure in a Ceph cluster at the same time could result in some data loss that otherwise could be avoided.

Reply:
I guess there is a case where the upgrade is still running AND the cluster is filling up quickly so we want to expand the cluster.